### PR TITLE
43 middleware chain

### PR DESCRIPTION
**Pull request by rsoletob**

### CODE REVIEW

File: middleware/middleware.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+// Middleware interface with all operations that must be implemented
+type Middleware interface {
+	// Action to do per each request
+	Action(next http.Handler) http.Handler
+
+	// Additional info that middleware use
+	GetInfo(r *http.Request, mc *MiddlewareContext)
+}
+
+// Handle method execute middlewares in correct order after API handler
```
Comment (adrianlop) [inline middleware/middleware.go]: after --> before

File: middleware/middleware.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+type MiddlewareContext struct {
+	// Authenticator middleware
+	UserId string
+	Admin  bool
+
+	// X-Request-Id middleware
+	XRequestId string
+}
+
+// Middleware interface with all operations that must be implemented
```
Comment (adrianlop) [inline middleware/middleware.go]: "Middleware interface with operations that all middlewares must implement"

File: middleware/middleware.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	XREQUESTID_MIDDLEWARE     = "XREQUESTID"
+	REQUEST_LOGGER_MIDDLEWARE = "REQUEST-LOGGER"
+)
+
+// MiddlewareHandler is handler with a middleware list to apply
+type MiddlewareHandler struct {
+	Middlewares map[string]Middleware
+}
+
+// MiddlewareContext with all parameters used in the context of middlewares
```
Comment (adrianlop) [inline middleware/middleware.go]: MiddlewareContext strcut contains all parameters used in the context of middlewares

File: middleware/middleware.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	REQUEST_ID_HEADER = "X-Request-Id"
+	USER_ID_HEADER    = "X-FOULKON-USER-ID"
+
+	// Middleware names
+	AUTHENTICATOR_MIDDLEWARE  = "AUTHENTICATOR"
+	XREQUESTID_MIDDLEWARE     = "XREQUESTID"
+	REQUEST_LOGGER_MIDDLEWARE = "REQUEST-LOGGER"
+)
+
+// MiddlewareHandler is handler with a middleware list to apply
```
Comment (adrianlop) [inline middleware/middleware.go]: MiddlewareHandler handles the HTTP request and applies its list of middlewares before calling the API

File: middleware/middleware.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	UserId string
+	Admin  bool
+
+	// X-Request-Id middleware
+	XRequestId string
+}
+
+// Middleware interface with all operations that must be implemented
+type Middleware interface {
+	// Action to do per each request
```
Comment (adrianlop) [inline middleware/middleware.go]: Action to apply to each request

**Comment by rsoletob** [comment]
Thanks for your comments @adrianlop, revamped.

**Pull request merged**